backend/cfg/cfg_invariants.ml

```diff
@@ -0,0 +1,176 @@
+ * of this software and associated documentation files (the "Software"), to deal  *
+ * in the Software without restriction, including without limitation the rights   *
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell      *
+ * copies of the Software, and to permit persons to whom the Software is          *
+ * furnished to do so, subject to the following conditions:                       *
+ *                                                                                *
+ * The above copyright notice and this permission notice shall be included in all *
+ * copies or substantial portions of the Software.                                *
+ *                                                                                *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR     *
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,       *
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE    *
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         *
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  *
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  *
+ * SOFTWARE.                                                                      *
+ *                                                                                *
+ **********************************************************************************)
+[@@@ocaml.warning "+a-30-40-41-42"]
+
+module CL = Cfg_with_layout
+
+type t =
+  { mutable result : bool;
+    ppf : Format.formatter;
+    fun_name : string;
+    cfg : Cfg.t;
+    mutable tailrec_entry_label : Label.t option
+  }
+
+let report t =
+  t.result <- false;
+  Format.fprintf t.ppf "Cfg invariant failed in %s: " t.fun_name;
+  Format.fprintf t.ppf
+
+let print_layout ppf layout =
+  Format.(pp_print_list ~pp_sep:pp_print_space pp_print_int ppf layout)
+
+let check_layout t layout =
+  (* [layout] is not empty and entry of cfg is the first node in the layout. *)
+  (match layout with
+  | [] -> report t "Empty layout"
+  | hd :: _ ->
+    if not (Label.equal hd t.cfg.entry_label)
+    then
+      report t "Cfg entry node %d is not the first first in the layout @.%a@."
+        t.cfg.entry_label print_layout layout);
+  (* No duplicates in layout *)
+  let labels =
+    List.fold_left
+      (fun acc l ->
+        if Label.Set.mem l acc then report t "Duplicate label %d in layout@." l;
+        Label.Set.add l acc)
+      Label.Set.empty layout
+  in
+  let num_labels = Label.Set.cardinal labels in
+  if List.compare_length_with layout num_labels > 0
+  then report t "Layout contains duplicates:@.%a@." print_layout layout;
+  (* The set of nodes in the layout is the same as the set of nodes in the
+     cfg. *)
+  let num_nodes = Label.Tbl.length t.cfg.blocks in
+  if num_nodes > num_labels
```

Comment:
It might be easier to read to simply
compute the set of labels from
`t.cfg.blocks`, and then perform two
set subtractions to show the labels
which are only on one side.